packages/core/src/combinator/applicative.js

```diff
@@ -16,5 +17,9 @@ import { apply } from '@most/prelude'
  * @returns {Stream} stream containing all the applications of fs to xs
  */
 export function ap (fs, xs) {
+  if (isCanonicalEmpty(xs)) {
```

Comment:
Couldn't this optimization be extended to `fs` as well?

Reply:
I was thinking the same but wasn’t sure.

Reply:
@TylorS would you return `fs` if canonical empty?

Reply:
Yes, I would.

Reply:
So two if-statements returning each a canonical empty. Another approach is one if-statement returning neither of the original streams but `empty()`. I think that would fit @briancavalier’s statements about that he doesn’t care if the return canonical empty is the same as the given.

Reply:
So two if-statements returning each a canonical empty. Another approach is one if-statement returning neither of the original streams but `empty()`. I think that would fit @briancavalier’s statements about that he doesn’t care if the return canonical empty is the same as the given.

Reply:
So two if-statements returning each a canonical empty. Another approach is one if-statement returning neither of the original streams but `empty()`. I think that would fit @briancavalier’s statements about that he doesn’t care if the return canonical empty is the same as the given.

Reply:
:+1:

Reply:
Because of types, I think specifically returning `empty()`, rather than returning either `xs` or `fs`, is the right thing (see [this comment](https://github.com/mostjs/core/pull/179#discussion_r152263187)).  So, I like the one `if`-statement idea: if `fs` or `xs` is canonical empty, return `empty()`.